_posts/2024-02-18-how-its-made.md

```diff
@@ -0,0 +1,180 @@
+The automated checks include what is called *continuous integration* (CI), which ensures the PR actually compiles on all platforms we support and not just the one the author is working on.
+
+![Pull request overview]({% link /static/img/post_2024-02-18-how-its-made/PR.png %}){:width="100%"}
+
+Automated checks are well and good, but so far they can't replace a human code review.
+In code review, other people look at the code and try to find any issues or spots that could be improved.
+This can be everything from validating functionality, checking UI and user experience, to proof-reading code comments.
+For a large change like the ship pathfinder, we've even organized a multi-player test game to check for any dreaded desync problems.
+
+Usually, this results in a back and forth between the author updating the PR and the reviewers looking again until the PR is either declared good or rejected.
+Unfortunately, not every author has the motivation to follow this until the end, so it happens that PRs just kind of fizzle off without reaching a finished state.
+Up to here, everybody can pitch in with reviewing a PR, but actually approving or rejecting a PR requires a decision by someone that has commit access to the OpenTTD repository.
+
+PR aren't actually outright rejected that often.
+Reasons for rejection might be that the PR is contrary to the goals of the OpenTTD project, like e.g. re-creating the original Transport Tycoon gameplay.
+Sometimes it also happens that a PR is meant well, but just not the right solution for a particular issue because it might affect some other part of the game negatively.
+
+For approving a PR we're quite strict, maybe stricter than many other projects out there.
+The OpenTTD project has a rigid code formatting style guide and commit message format.
+We're also unlikely to accept a PR with known problems or unfinished parts.
+For example, the game is translated to many languages, including some that are written right-to-left, and any UI code needs to be coded to work for all languages and not just English.
+Finally, GitHub will prevent approving a PR if any of the CI checks have failed.
+
+While all this might sometimes appear to be overly pedantic and petty, a strict quality control is important to make sure the game can continue for the next 20 years.
+Everybody has probably noticed that everything, from your desk to the storage shelf in the shed, seems to always descend into chaos unless cleaned up regularly, and code is no exception to this.
+If we wouldn't care about code quality, the source code would probably be a hot mess after 20 years and full of bugs, making any new features really difficult.
+
+One thing of note here is that no author can self-approve a PR, even if they have commit access to OpenTTD.
+In this sense, all PR authors are treated equal and need to follow the same process to get a PR approved.
+
+Finally, if a PR has the coveted approval, it is merged into the main code base.
```

Comment:
This might be too much detail, but sometimes contributors wonder why their approved PR hasn't been merged yet.

```suggestion
Finally, if a PR has the coveted approval, another developer can merge it into the main code base. Like a nuclear missile silo, we usually wait for two maintainers to turn their keys: one to approve, a second to merge. If the PR was opened by someone with commit access, they can merge their own PR after it is approved.
```

Reply:
Well, I've expanded this a bit, even if I didn't used your suggestion directly.